### ISSUE TITLE
Add support for functional custom messages

### ISSUE THREAD

**Issue by amk221** [stylelint/stylelint#4117]
> What is the problem you're trying to solve?

I would like to inform developers better when their code fails linting.

Currently, I can provide a custom message but:

1. This loses the ability to reference the _value_ that caused the failure. For example:
   * Before: `Unexpected hex color "#fff"`
   * After `Unexpected hex color. Please refer to the brand colours in the Styleguide`
<br><br>
2. The blacklisted values are a great way to prevent developers from adding rules you would otherwise not want. However, it's not possible to do this:
    ```
    'declaration-property-value-blacklist': {
       'font-weight': ['500', { message: 'Please use the semi-bold mixin instead' }]
    },
    ```
    ...you can only add 1 catch-all message for _all_ invalid property values.
<br><br>
> What solution would you like to see?

1.  It would be great to be able to swap-in values `${hex}`
     or, at least prefix/suffix the original error to the custom message.

2. It would be nice if the custom messages were acknowledged for each invalid property

**Comment by hudochenkov** [comment]
This are two different suggestions.

1. Allow custom message to be a function. It's a nice idea, but could be rather complex. As every rule has different values it's provides to default message. Most rules has zero or one, but there are some rules which expose two or more. How to communicate how many and what values could be in custom message is a big problem.

2. Allow extra description to each value in `declaration-property-value-blacklist`. While this one sounds easier, it could be complicated as well. Because it's not property-value pairs, it's property-values pairs.

	This is how config might look like:

	```js
	{
		rules: {
			'declaration-property-value-blacklist': {
				'font-weight': [
					['400', { message: 'Please use the normal mixin instead' }],
					['500', { message: 'Please use the semi-bold mixin instead' }],
					['700', { message: 'Please use the bold mixin instead' }],
				],
				"transform": [
					[
						['/scale3d/', '/rotate3d/', '/translate3d/'],
						{ message: 'Do not use 3D transformation' }
					],
					[
						['/skew/'],
						{ message: 'Do not use skew' }
					],
				],
			},
		}
	}
	```

	As you see it would be rather confusing to configure. And very error-prone.

**Comment by amk221** [comment]
Ok thanks for the info.

Would it be possible to make `message` a function? 

```javascript
'declaration-property-value-blacklist': [{
   'font-weight': ['500']
}, {
  message: function(prop, value) {}
}],
```

Then the onus is on me to do the work, not Stylelint

Mind you, that gets us into custom formatter territory.

**Comment by kwokrobyn1** [comment]
Any updates on this? I noticed it was closed, but did not see any changes being made

**Comment by glen-84** [comment]
@amk221 Can you please reopen this issue?

Not being able to specify a custom message for a blacklisted property value is a problem.

**Comment by tmoffett-centrify** [comment]
I also would like to be able to provide a custom message while still having the value, which caused the failure, reported, as stated by @amk221 for issue #1 in the original post.

**Comment by Mouvedia** [comment]
Can we reopen this for `declaration-property-value-disallowed-list` rule?

**Comment by glen-84** [comment]
@Mouvedia It ... is open ... ?

**Comment by ybiquitous** [comment]
I've created PR #6312 as a PoC.
